### DISCUSSION TITLE
Add utility for `scipy.linalg.qr_update`

### DISCUSSION THREAD

**Discussion by YigitElma** [jax-ml/jax#23149] [Ideas]
For my application, I need to take QR of 

$$
\tilde{Q} \tilde{R} = \begin{pmatrix} J \\ 
\alpha I \end{pmatrix}
$$

where I already have

$$
QR = J
$$

So, I have tried to implement this myself using Givens rotations. But due to `N*M` loops it is very slow, which makes it useless since the main goal of this function is to calculate QR faster than otherwise calling `jax.scipy.linalg.qr`. I guess this would be useful for people using `jax` for trust region Newton method optimization. I thought of using Householder reflections but I couldn't implement it under `jax.jit`. For reference, this is implemented in `scipy` [here](https://github.com/scipy/scipy/blob/main/scipy/linalg/_decomp_update.pyx.in) and calls some LAPACK routines in a for loop. I hope this can be implemented in the same way some other matrix decomposition routines are implemented in JAX.

I naively tried to implement this myself purely in Python like this,
```python
def _givens_jax(a, b):
    b_zero = abs(b) == 0
    a_lt_b = abs(a) < abs(b)
    t = -jnp.where(a_lt_b, a, b) / jnp.where(a_lt_b, b, a)
    r = rsqrt(1 + abs(t) ** 2).astype(t.dtype)
    cs = jnp.where(b_zero, 1, jnp.where(a_lt_b, r * t, r))
    sn = jnp.where(b_zero, 0, jnp.where(a_lt_b, r, r * t))
    G2 = jnp.array([[cs, -sn], [sn, cs]])
    return G2.astype(float)

@jax.jit
def update_qr_jax(A, w, q, r):
    """Update QR factorization with a diagonal matrix w at the bottom."""
    m, n = A.shape
    Q = jnp.eye(m + n)
    Q = Q.at[:m, :m].set(q)

    R = jnp.vstack([r, w])

    def body_inner(i, jQR):
        j, Q, R = jQR
        i = m + j - i
        a, b = R[i - 1, j], R[i, j]
        G2 = _givens_jax(a, b)
        R = R.at[jnp.array([i - 1, i])].set(G2 @ R[jnp.array([i - 1, i])])
        Q = Q.at[:, jnp.array([i - 1, i])].set(Q[:, jnp.array([i - 1, i])] @ G2.T)
        return j, Q, R

    def body(j, QR):
        Q, R = QR
        j, Q, R = fori_loop(0, m, body_inner, (j, Q, R))
        return Q, R

    Q, R = fori_loop(0, n, body, (Q, R))
    R = jnp.where(jnp.abs(R) < 1e-10, 0, R)

    return Q, R
```
Note: I also tried `economic` mode QR to reduce matrix size, but this is still slow.

**Comment by YigitElma** [comment]
@patrick-kidger @jpbrodrick89, maybe you might have some idea on this? Does such a function have an equivalent in `lineax` or is there a way to solve trust region problems efficiently? To be more specific, what I am solving is explained [here](https://desc-docs.readthedocs.io/en/stable/dev_guide/notebooks/getting-started-eq-solve.html#Solving-Trust-Region-Problem-with-QR)

**Reply by patrick-kidger** [reply]
I recognise this problem as being the kind of one that comes up a lot in e.g. Levenberg Marquardt. When we implemented this then we decided to just call a fresh QR each time, mostly for reasons of simplicity (we had other problems to tackle rather than writing out Givens/Householder).

...Which is to say, sorry, we don't have an implementation of this either.

**Reply by jpbrodrick89** [reply]
Thanks for reaching out @YigitElma. It's great to be aware of the real world use cases of least squares especially as I try envisage what future capabilities might be reasonable or not for lineax (for example I've recently been exploring least squares with linear equality constraints which has some crossover). For your problem specifically, I just opened a [PR](https://github.com/PlasmaControl/DESC/pull/2244) in DESC for a simple suggestion using the new [qr_multiply](https://docs.jax.dev/en/latest/_autosummary/jax.scipy.linalg.qr_multiply.html). This is not ALL the improvement you want but on my benchmarks I got 2-4x speedup which might be enough for you to beat SVD even at the beginning of your nonlinear solve.

For posteriority here is the relevant part of the PR comment:

Reuse QR factorisation of undamped Jacobian, this reduces the computational cost of each alpha-solve by a factor (m+n)/2n (where height m > n) at the cost of one additional extra initial solve:

$$
\begin{pmatrix}
J \\
\sqrt{a} I
\end{pmatrix}∆x =
\begin{pmatrix}
Q_0 \begin{pmatrix}
R_0 \\ 
0 \end{pmatrix}\\
\sqrt{a} I
\end{pmatrix}
$$

$$
\implies
\begin{pmatrix}
Q_0^{(\text{thin})T} & 0 \\
0 & I 
\end{pmatrix} 
\begin{pmatrix}
J \\
\sqrt{a} I
\end{pmatrix}=
\begin{pmatrix}
R_0 \\
\sqrt{a} I
\end{pmatrix} = 
Q_1 \begin{pmatrix}
R_1 \\
0
\end{pmatrix}
$$

So all one needs to do is a QR of the 2n x n matrix formed by vertically stacking $R_0$ and $\sqrt{alpha}I$. Alternatively, you could view this as Cholesky update and either scan over [jax.lax.linalg.cholesky_update](https://docs.jax.dev/en/latest/_autosummary/jax.lax.linalg.cholesky_update.html) (stable but likely as slow as your homebrewed version) or form the Gram matrix $R_0^T R_0 + \alpha I$ which is probably the fastest option but as always will struggle with ill-conditioning. What you ideally want is a wrapper for something like [Lapack](https://www.netlib.org/lapack//explore-html/d4/dad/group__tpqrt_gadb01a522e74a81897506c27a3508e35c.html#gadb01a522e74a81897506c27a3508e35c)'s `dtpqrt` which is specifically designed for factorising a stacked triangular plus trapezoidal matrix but I'm not sure if there's a GPU equivalent.

@patrick-kidger I think we're a way off from being ready for this, but one could envisage supporting this functionality with something like `solver.pivot` and `solver.update` in lineax and then threading it into optimistix's `IndirectLevenbergMarquadt`. Obviously, getting the design right to avoid overcomplexity and API bloat is important and not something to jump into blindfolded.

**Reply by YigitElma** [reply]
Thank you very much for the quick reply! And special thanks for the PR 🙏 I think we can continue on the discussion in the [DESC PR](https://github.com/PlasmaControl/DESC/pull/2244)
